ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java

```diff
@@ -29,38 +30,87 @@
+    this.adminClientSupplier = new NotThreadSafeMemoizedSupplier<Admin>(adminClientSupplier);
+
+    requireNonNull(srClientSupplier, "srClientSupplier");
+    this.srClientSupplier = new NotThreadSafeMemoizedSupplier<SchemaRegistryClient>(
+        srClientSupplier);
+
+    requireNonNull(connectClientSupplier, "connectClientSupplier");
+    this.connectClientSupplier = new NotThreadSafeMemoizedSupplier<ConnectClient>(
+        connectClientSupplier);
+
+    requireNonNull(ksqlClientSupplier, "ksqlClientSupplier");
+    this.ksqlClientSupplier = new NotThreadSafeMemoizedSupplier<SimpleKsqlClient>(
+        ksqlClientSupplier);
+
+    this.srClient = requireNonNull(srClientSupplier.get(), "srClient");
+
+    this.kafkaClientSupplier = requireNonNull(kafkaClientSupplier, "kafkaClientSupplier");
+
+    this.topicClientSupplier = new NotThreadSafeMemoizedSupplier<KafkaTopicClient>(
+        () -> new KafkaTopicClientImpl(this.adminClientSupplier));
+  }
+
+  @VisibleForTesting
+  public DefaultServiceContext(
+      final KafkaClientSupplier kafkaClientSupplier,
+      final Supplier<Admin> adminClientSupplier,
       final KafkaTopicClient topicClient,
-      final Supplier<SchemaRegistryClient> srClientFactory,
-      final ConnectClient connectClient,
-      final SimpleKsqlClient ksqlClient
+      final Supplier<SchemaRegistryClient> srClientSupplier,
+      final Supplier<ConnectClient> connectClientSupplier,
+      final Supplier<SimpleKsqlClient> ksqlClientSupplier
   ) {
+
+    requireNonNull(adminClientSupplier, "adminClientSupplier");
+    this.adminClientSupplier = new NotThreadSafeMemoizedSupplier<Admin>(adminClientSupplier);
+
+    requireNonNull(srClientSupplier, "srClientSupplier");
+    this.srClientSupplier = new NotThreadSafeMemoizedSupplier<SchemaRegistryClient>(
+        srClientSupplier);
+
+    requireNonNull(connectClientSupplier, "connectClientSupplier");
+    this.connectClientSupplier = new NotThreadSafeMemoizedSupplier<ConnectClient>(
+        connectClientSupplier);
+
+    requireNonNull(ksqlClientSupplier, "ksqlClientSupplier");
+    this.ksqlClientSupplier = new NotThreadSafeMemoizedSupplier<SimpleKsqlClient>(
+        ksqlClientSupplier);
+
+    this.srClient = requireNonNull(srClientSupplier.get(), "srClient");
+
     this.kafkaClientSupplier = requireNonNull(kafkaClientSupplier, "kafkaClientSupplier");
-    this.adminClient = requireNonNull(adminClient, "adminClient");
-    this.topicClient = requireNonNull(topicClient, "topicClient");
-    this.srClientFactory = requireNonNull(srClientFactory, "srClientFactory");
-    this.srClient = requireNonNull(srClientFactory.get(), "srClient");
-    this.connectClient = requireNonNull(connectClient, "connectClient");
-    this.ksqlClient = requireNonNull(ksqlClient, "ksqlClient");
+
+    requireNonNull(topicClient, "topicClient");
+    this.topicClientSupplier = new NotThreadSafeMemoizedSupplier<KafkaTopicClient>(
```

Comment:
nit: this constructor should just call the other one

Reply:
+1

Reply:
+1.

Though this is a little tricky. I'll make the change and post in my review.